lib/prefetch.js

```diff
@@ -0,0 +1,113 @@
+import React from 'react'
+import Link, { isLocal } from './link'
+import { parse as urlParse } from 'url'
+
+class Messenger {
+  constructor () {
+    this.id = 0
+    this.callacks = {}
+    this.serviceWorkerReadyCallbacks = []
+    navigator.serviceWorker.addEventListener('message', ({ data }) => {
+      if (data.action !== 'REPLY') return
+      if (this.callacks[data.replyFor]) {
+        this.callacks[data.replyFor](data)
+      }
+    })
+    this._register()
+  }
+
+  send (payload, cb) {
+    if (this.serviceWorkerState === 'REGISTERED') {
+      this._send(payload, cb)
+    } else {
+      this.serviceWorkerReadyCallbacks.push(() => {
+        this._send(payload, cb)
+      })
+    }
+  }
+
+  _send (payload, cb = () => {}) {
+    const id = this.id ++
+    const newPayload = { ...payload, id }
+
+    this.callacks[id] = (data) => {
+      if (data.error) {
+        cb(data.error)
+      } else {
+        cb(null, data.result)
+      }
+
+      delete this.callacks[id]
+    }
+
+    navigator.serviceWorker.controller.postMessage(newPayload)
+  }
+
+  _register () {
+    this.serviceWorkerState = 'REGISTERING'
+    navigator.serviceWorker.register('/_next-prefetcher.js')
+
+    // Reset the cache after every page load.
+    // We don't need to have any old caches since service workers lives beyond
+    // life time of the webpage.
+    // With this prefetching won't work 100% if multiple pages of the same app
+    // loads in the same browser in same time.
+    // Basically, cache will only have prefetched resourses for the last loaded
+    // page of a given app.
+    // We could mitigate this, when we add a hash to a every file we fetch.
+    const reset = () => {
+      this._send({ action: 'RESET' }, (err) => {
+        if (err) throw err
+        this.serviceWorkerState = 'REGISTERED'
+        this.serviceWorkerReadyCallbacks.forEach(cb => cb())
+        this.serviceWorkerReadyCallbacks = []
+        console.log('Next Prefetcher registered')
+      })
+    }
+
+    if (navigator.serviceWorker.controller) {
+      reset()
+    } else {
+      navigator.serviceWorker.oncontrollerchange = reset
+    }
+  }
+}
+
+function hasServiceWorkerSupport () {
+  return (typeof navigator !== 'undefined' && navigator.serviceWorker)
+}
+
+const PREFETCHED_URLS = {}
+let messenger
+
+if (hasServiceWorkerSupport()) {
+  messenger = new Messenger()
+}
+
+export function prefetch (href) {
+  if (!hasServiceWorkerSupport()) return
+  if (!isLocal(href)) return
+
+  let { pathname } = urlParse(href)
+  // Add support for the index page
+  if (pathname === '/') {
+    pathname = '/index'
+  }
+
+  const url = `${pathname}.json`
```

Comment:
Btw this url will become the same with non-json url and need to set the `Accept` header with https://github.com/zeit/next.js/pull/310